mesh/reward.go

```diff
@@ -0,0 +1,100 @@
+package mesh
+
+import (
+	"github.com/spacemeshos/go-spacemesh/common"
+	"github.com/spacemeshos/go-spacemesh/log"
+	"github.com/spacemeshos/go-spacemesh/state"
+	"math/big"
+)
+
+type RewardParams struct {
+	SimpleTxCost   *big.Int
+	BaseReward     *big.Int
+	PenaltyPercent *big.Int
+	TxQuota        uint32
+}
+
+//type Transactions []*state.Transaction
+
+func MergeDoubles(transactions []*state.Transaction) []*state.Transaction {
+	transactionSet := make(map[common.Hash]struct{})
+	merged := make([]*state.Transaction, 0, len(transactions))
+	for _, trns := range transactions {
+		if _, ok := transactionSet[trns.Hash()]; !ok {
+			transactionSet[trns.Hash()] = struct{}{}
+			merged = append(merged, trns)
+		} else {
+			log.Info("double trans merged %v", trns)
+		}
+	}
+	return merged
+}
+
+func (m *Mesh) CalculateLayerReward(id LayerID, params RewardParams) *big.Int {
+	return params.BaseReward
+}
+
+func (m *Mesh) AccumulateRewards(rewardLayer LayerID, params RewardParams) {
+	l, err := m.getLayer(rewardLayer)
+	if err != nil || l == nil {
+		m.Error("") //todo handle error
+		return
+	}
+
+	ids := make(map[string]struct{})
+	uq := make(map[string]struct{})
+
+	//todo: check if block producer was eligible?
+	for _, bl := range l.blocks {
+		if _, found := ids[bl.MinerID]; found {
+			log.Error("two blocks found from same miner %v in layer %v", bl.MinerID, bl.LayerIndex)
+			continue
+		}
+		ids[bl.MinerID] = struct{}{}
+		if uint32(len(bl.Txs)) < params.TxQuota {
+			//todo: think of giving out reward for unique txs as well
+			uq[bl.MinerID] = struct{}{}
+		}
+	}
+	//accumulate all blocks rewards
+	txs := m.ExtractOrderedSortedTransactions(l)
```

Comment:
putting struct methods in multiple files is not the normal convention 
can we move this or change the reciver ?